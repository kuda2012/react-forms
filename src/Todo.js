import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Item from "./Item";
import NewItemForm from "./NewItemForm";

const Todo = () => {
  const [items, setItems] = useState([]);
  const addItem = (newItem) => {
    setItems((items) => [...items, { ...newItem, id: uuid() }]);
  };
  const deleteItem = (id) => {
    setItems((items) => {
      return items.filter((item) => item.id !== id);
    });
  };
  return (
    <>
      <h3>To-do List</h3>
      <NewItemForm addItem={addItem} />
      {items.length !== 0 && (
        <ul className="Todo" data-testid="Todo">
          {items.map((item, i) => (
            <Item
              key={item.id}
              id={item.id}
              todo={item.todo}
              number={i}
              deleteItem={deleteItem}
            />
          ))}
        </ul>
      )}
    </>
  );
};

export default Todo;
