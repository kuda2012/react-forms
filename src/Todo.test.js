import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Todo from "./Todo";

it("renders without crashing", function () {
  render(<Todo />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<Todo />);
  expect(asFragment()).toMatchSnapshot();
});
it("shows title", function () {
  const { queryByText } = render(<Todo />);
  const title = queryByText("To-do List");
  expect(title).toBeInTheDocument();
});
it("should make a new Todo", () => {
  const { queryByLabelText, queryByText, queryByTestId } = render(<Todo />);
  const TodoListBeforeCreation = queryByTestId("Todo");
  expect(TodoListBeforeCreation).not.toBeInTheDocument();
  const todo = queryByLabelText("Enter To-do Item");
  const btn = queryByText("Add Item");
  fireEvent.change(todo, { target: { value: "Kuda" } });
  fireEvent.click(btn);
  const TodoListAfter = queryByTestId("Todo");
  expect(TodoListAfter).toBeInTheDocument();
  const itemCreated = queryByTestId("Item0");
  expect(itemCreated).toBeInTheDocument();
});

it("should delete a todo", () => {
  const { queryByLabelText, queryByText, queryByTestId } = render(<Todo />);
  const TodoListBeforeCreation = queryByTestId("Todo");
  expect(TodoListBeforeCreation).not.toBeInTheDocument();
  const todo = queryByLabelText("Enter To-do Item");
  const btn = queryByText("Add Item");
  fireEvent.change(todo, { target: { value: "Kuda" } });
  fireEvent.click(btn);
  const TodoListAfter = queryByTestId("Todo");
  expect(TodoListAfter).toBeInTheDocument();
  const itemCreated = queryByTestId("Item0");
  expect(itemCreated).toBeInTheDocument();
  const deleteItem = queryByText("X");
  fireEvent.click(deleteItem);
  expect(TodoListAfter).not.toBeInTheDocument();
  expect(itemCreated).not.toBeInTheDocument();
});
