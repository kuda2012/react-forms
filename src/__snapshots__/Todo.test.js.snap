// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h3>
    To-do List
  </h3>
  <form>
    <label
      for="todo"
    >
      Enter To-do Item
    </label>
    <input
      id="todo"
      name="todo"
      type="text"
      value=""
    />
    <button>
      Add Item
    </button>
  </form>
</DocumentFragment>
`;
